docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,185 @@
+---
+title: Resource Pinning
+description: Protection of resources from accidental deletion/moves
+---
+
+# Resource Pinning
+
+- **Author(s):**: @MarkMcCulloh
+- **Submission Date**: 2024-02-15
+- **Stage**: Proposal
+- **Stage Date**: 2024-02-15
+
+Introduce a mechanism to protect resources from accidental deletion/moves in the logical application tree.
+
+## Requirements
+
+### Functional
+- User-ability in sources to signify intent for a resource to not be deleted or moved (e.g. "pinning")
+- Git-trackable way to store changes to pinned resources between compilations, per target platform (e.g. "pinfile")
+- User-ability to manually and declaratively override a pinned resource's path in the resource tree
+- Compilation failure when a pinned resource is deleted without explicit removal from the pinfile
+- Compilation failure when a pinned resource is moved without explicit mapping of a previous path to a new path
+- Compilation failure when a resource is no longer pinned in source but is still present in the pinfile
```

Comment:
Resource replacement can be a result of logical id changes but *also* a result of changing certain attributes. This is engine+resource dependent.

I think we need to at least have an idea for how this interplays with the replacement behavior of the underlying provisioning engine.

For example, I would expect (maybe unjustifiably) that if I pin a resource then Terraform or CFN will never replace it, so if I modify an attribute that causes replacement, I would get an error as early as possible.

Reply:
Sure I think this makes sense and I kinda mention this in the out-of-scope section. I think the target can choose its own behavior here. The target itself knows which things are pinned, so it can do this sort of introspection/validation.

Should this be part of this RFC?

Reply:
I like the suggestion to include this in the pinfile as an attribute. Let's mention this in the RFC.

Reply:
> For example, I would expect (maybe unjustifiably) that if I pin a resource then Terraform or CFN will never replace it, so if I modify an attribute that causes replacement, I would get an error as early as possible.

After thinking on this for a while I realized that this error still can't happen at compile time because we can't reliable store a real value for arbitrary attributes.

Consider the example of a bucket. The `name` attribute is likely something you'd want to store to ensure it doesn't changed, because that would cause a replacement. However, the name of a bucket is not always statically known because it could contain something like a token. So there would be nothing to pin without referencing a statefile or somehow hook into deployment.

There are two ways that pinning still overall solves the problem:

1. Provisioning features like `prevent_destroy` aren't perfect, but do totally fix this scenario because it's only a problem with mutation rather than resource orphaning. This means this would still be caught before deployment finishes at the very least. The platform can take care to add this (or equivalent) attribute to all pinned resources.
3. I added a section to the pinfile called `attributes`. This allows you to override the raw platform-level data for a resource. Instead of using this to catch errors, this serves as a way to override values.

Reply:
Yeah, it feels to me like knowing whether a set of attributes will result in a resource deletion / recreation is a turing-complete problem (it depends on how the underlying resource provider is implemented) so it might be worth leaving these kinds of edge cases to the provisioning engine.

For example, I could change a piece of Terraform config from `prop: "MyBucket12345-Backup" to `prop: "${other_bucket.bucket_name}-Backup" (or change my Wing code to cause that change), and it might _not_ result in a replacement since the Terraform provider can figure out that they resolve to the same string during `terraform plan` / `terraform deploy`. But there's no way the Wing compiler would be able to figure that out.

Providing an escape hatch where the Wing platform provider can include some extra information (a hash, or JSON object) in the pinfile for detecting sensitive changes could be worthwhile sounds ok (though I'd also be okay leaving it out if it seems like a capability we can add on later easily).